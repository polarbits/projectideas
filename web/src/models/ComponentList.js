import {types} from "mobx-state-tree";

export const component = {
  "type": 'textarea',
  "content": 'This is first task todo in this project'
}

export const ComponentListItem = types.model({
  type: types.string,
  content: types.string
})
.actions(self => ({
  changeType(newType){
    self.type = newType
  },
  changeContent(newContent){
    self.content = newContent
  },
}))
  
  // {
  // function setType(newType) {
  //   self.type = newType
  // }
  // return {setType}
// })

export const ComponentList = types.model({
  items: types.optional( types.array(ComponentListItem),[])
})
.actions(self => ({
  add(item) {
    self.items.push(item)
  },
  changeContent(newContent) {
    self.content = newContent;
  }
}))

