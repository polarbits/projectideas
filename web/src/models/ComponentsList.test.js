import {getSnapshot, onSnapshot, onPatch} from "mobx-state-tree"
import {ComponentList,ComponentListItem} from "./ComponentList"

it("can create a instance of a model", () => {
  const item = ComponentListItem.create({
    type: 'textarea',
    content: 'This is first task todo in this project'
  });
  expect(item.type).toBe("textarea");
  expect(item.content).toBe("This is first task todo in this project");
  item.changeType("image")
  expect(item.type).toBe("image");
})

it("can create a component list", () => {
  const list = ComponentList.create({
    items: [
      {
        type: 'textarea',
        content: 'This is first task todo in this project'
      }
    ]
  });
  expect(list.items.length).toBe(1)
})

it('can add new item(component)',() => {
  const list = ComponentList.create()
  list.add({
    type: "textarea",
    content: "test text content"
  })
  const states = [];
  onSnapshot(list, snapshot => {
    states.push(snapshot);
  })

  expect(list.items.length).toBe(1)
  expect(list.items[0].type).toBe("textarea")
  list.items[0].changeContent("new content")
  expect(list.items[0].content).toBe("new content")

  expect(getSnapshot(list)).toEqual({
    items:[
      {
        type: "textarea",
        content: "new content"
      }
    ]
  })

  expect(getSnapshot(list)).toMatchSnapshot()
})

it('can add new item(component) - 2',() => {
  const list = ComponentList.create()
  list.add({
    type: "textarea",
    content: "test text content"
  })

  const patches = [];
  onPatch(list, patch => {
    patches.push(patch);
  })

  list.items[0].changeContent("new content")

  expect(patches).toMatchSnapshot();
})