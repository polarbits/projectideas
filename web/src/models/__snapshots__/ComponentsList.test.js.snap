// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new item(component) - 2 1`] = `
Array [
  Object {
    "op": "replace",
    "path": "/items/0/content",
    "value": "new content",
  },
]
`;

exports[`can add new item(component) 1`] = `
Object {
  "items": Array [
    Object {
      "content": "new content",
      "type": "textarea",
    },
  ],
}
`;
